### PR TITLE
feat: consistent unique route names for Ingress, HTTPRoute

### LINKED ISSUE #834: Route naming is not stable
The route naming scheme we use currently (for v1beta1, v1, tcp and knative ingress) is not stable:

```go
Route: kong.Route{                                                                                                                                                                                                                                                      
  // TODO (#834) Figure out a way to name the routes                                                                                                                                                                                                                           
  // This is not a stable scheme                                                                                                                                                                                                                                        
  // 1. If a user adds a route in the middle,                                                                                                                                                                                                                           
  // due to a shift, all the following routes will                                                                                                                                                                                                                      
  // be PATCHED                                                                                                                                                                                                                                                         
  // 2. Is it guaranteed that the order is stable?                                                                                                                                                                                                                      
  // Meaning, the routes will always appear in the same                                                                                                                                                                                                                 
  // order?                                                                                                                                                                                                                                                             
  Name:          kong.String(ingress.Namespace + "." + ingress.Name + "." + strconv.Itoa(i) + strconv.Itoa(j)),                                                                                                                                                         
  ...
}
```

The scope of this issue is to fix the issue identified in the comment above.

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

This adds consistent, unique names for `kong.Routes` derived from HTTPRoute rules.

**Which issue this PR fixes**

Resolves #834 
Resolves https://github.com/Kong/kubernetes-ingress-controller/issues/2156

**Important Notes**

This adds the hashed route name features as originally desired in #834, but in order to preserve backwards compatibility (mostly with metrics) it adds them under the new `--hashed-dataplane-route-names` flag which is disabled by default. The flag has been added to anonymous reports so we can understand whether people are enabling this in the wild over time.

I considered doing a larger refactoring of `internal/dataplane/parser` in order to accommodate data-plane related options, but given the size of the PR already and without having a KEP and some consensus among the team on how that should look I decided to go with the approach of simply adding on the option as a parameter down the function call chain. We can revisit this later as part of our #2209 spike.

I added a new CI test run to test "extended options" with the comment that this is a test for any options or flags on the manager which have wide ranging impacts on multiple components this enables us to run all integration tests with and without `--hashed-dataplane-route-names` enabled since this feature has far reaching effects.

**Checklist**

- [x] the CHANGELOG.md has been updated

### CODE REVIEW

**Review by rainest — COMMENTED** [review]
Big enough that I wanna sleep on it and do a second read. Initial comment is the one thing that jumped out as weird immediately.

File: internal/controllers/utils/ingress_status.go [hunk 1]
```diff
@@ -704,3 +686,64 @@ func lessLoadBalancerIngress(addrs []apiv1.LoadBalancerIngress) func(int, int) b
...
+		if kubernetesVersion.Major >= uint64(1) && kubernetesVersion.Minor > uint64(18) {
+			if err := UpdateIngress(ctx, log, nsn, client, ips); err != nil {
+				return fmt.Errorf("failed to update ingressv1: %w", err)
+			}
+		} else {
+			// TODO: this can go away once we drop support for Kubernetes older than v1.19
+			if err := UpdateIngressLegacy(ctx, log, nsn, client, ips); err != nil {
+				return fmt.Errorf("failed to update ingressv1: %w", err)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/utils/ingress_status.go:745]: Actually this was a miss and needs to be added back. Also looks like some kind of integration test needs to be added to catch this.

[the remark this answers is not in the document]
Comment (rainest) [reply internal/controllers/utils/ingress_status.go:745]: Odd, I'd have thought https://github.com/Kong/kubernetes-ingress-controller/blob/a52e3ff6ed9bcfbe13d19595553e62433394cf79/test/integration/udpingress_test.go#L132-L147 would have caught it already.

[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/utils/ingress_status.go:745]: It was the end of the day when I responded and it just looked like a miss, but this morning after digging back in: I accounted for this in the `isRouteUsingLegacyNaming()` function: https://github.com/Kong/kubernetes-ingress-controller/pull/2222/files#diff-49a079b80173f1f365649af762f004c2c74e0d8daca714c2d44e55a59a29e871R149

Basically when we released UDP we added a suffix that allowed it to be uniquely identified as a `UDPIngress` resource (and the same is true for `HTTPRoute`, though its unreleased) so this one doesn't need to use the fallback code. I wrote some comments which point out which types of objects will need to use the fallback code: https://github.com/Kong/kubernetes-ingress-controller/pull/2222/files#diff-0d82b90e9114ea2d9714794ab0077aca4ec5d5c4c3b6ba0a0c4a6e6612388b74R240

So my bad this was accounted for this PR had just taken so long it had been a minute.

**Review by rainest — COMMENTED** [review]
The actual code here looks fine from a "does what it says" perspective, and I don't think I have anything important to add to that, so this is more commentary on the change itself and questions on whether we could gate this more neatly in advance of the discussion later this week:

- Simplification of the status update system by making each route name affirmatively declare what the parent K8S kind was is good. Having to continue to maintain the legacy stuff makes that less of a win though, and we end up with additional spaghetti due to stuff like https://github.com/Kong/kubernetes-ingress-controller/pull/2222/files#r798590532, where sometimes the legacy behavior is handled in one place, and other times in another. We should consider keeping this without the flag.
- There are two practical use cases for route names: querying the admin API and Prometheus/Vitals metrics. Hashes break the former entirely, as they're not human-predictable. Rule indices may change, but they always reflect the current state of an Ingress using a simple pattern. Users can construct those names by hand easily. Metrics are in a weird space because while we want continuity, you don't really get continuity with hashing unless you never change any configuration, only add additional rules. Continuity by name is furthermore confusing because [Kubernetes users don't interact with our generated names, they interact with Kubernetes objects](https://discuss.konghq.com/t/dealing-with-the-lack-of-kubernetes-in-kong-metric-labels/9604). Metrics for the objects are more useful, but we lack a good way to provide that metadata. Ideally Kong entities would have Kubernetes metadata to indicate the parent object, and we should push for a proper Gateway-side way to insert these. We do have a lazy option of abusing tags, and setting something like "k8sKind:Ingress,k8sName:foo,k8sNamespace:bar" and then adding code to Vitals/Prometheus to slurp that into the appropriate tag. Kinda ugly, but tags have the benefit of being ubiquitous, and we shouldn't need anything more complex than a single string.
- In absence of an actual user-declared name/ID for individual rules, there is no true "consistent" scheme, only "invariant to X", because we're always using something not intended for use as a name to derive a name. If you really want consistent names, GWAPI route kinds have these (maybe?).
- This does work well as a case study in what does or does not work well for feature gates: if most of the new code is dead without a high-level on/off switch (Gateway, by virtue of the controller setup populating or not populating any resources for the always-on parser code to consume), feature gates work well. If something needs to be plumbed through multiple layers of functions, and will require separate test sets, not so well. I think this particular change is, in retrospect, a poor fit for gates in the best of scenarios, but I am curious if there's a design pattern we can use that allows us to put the inflection point further down than `manager/setup.go` without passing a `bool` through umpteen function signatures?

**Comment by shaneutt** [comment]
https://github.com/Kong/kubernetes-ingress-controller/pull/2264 follows up to remove the TODOs.

**Pull request closed without merging**